text/3738-drop-type-destructuring.md

```diff
@@ -0,0 +1,392 @@
+- `$e` must be an owned expression that can be matched by pattern `$pat`
+- `$e`'s type must be an ADT (`struct` or `enum`)
+- The type of `$e` must be constructable in the current context, i.e.
+    - All fields must be visible to the current module wrt privacy
+    - If the type is marked with `#[non_exhaustive]`, it must be defined in the current crate
+
+The semantic of this macro is equivalent to the following:
+1. Wrapping `$e` in a `ManuallyDrop`
+2. Copying **all** fields of `$e` (for enums all fields of the active variant)
+3. If the `let` token is present in the macro invocation, fields that are listed in `$pat` are binded to their respective bindings or patterns. Otherwise, they are assigned similarly to destructuring assignment
+4. If `..` is present in `$pat`, all fields that are not mentioned in it are dropped in the order of definition (simialrly to how [normal `let`/destructuring assignment works](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=3abd3aebd3378690ff3d2006e12d4120))
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+When implementing this change as a macro expanding to a statement, we do not think there are many drawbacks. However, in the coming section with alternatives we discuss another possible expansion to a pattern which may have a few advantages, but also has many drawbacks which we discuss there.
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+## Do not support destructuring assignment
+
+Instead of allowing optional `let` token, we could always ommit it, and consider `destructure!` to always introduce new bindings.
+
+It's not evident that destructure assignment is particularly useful in situations where `destructure!` would be used.
+
+## Expand to a pattern
+
+Instead of making this macro expand to a statement, containing a pattern and a right hand side, one could imagine a macro that expands *just* to a pattern:
+
+```rust
+let destructure!(Foo { a, b }) = x;
+```
+
+On its own this could be considered, it also means that whether to use `let` or not (the previous alternative) isn't a concern for the macro. However, by doing this, we might imply that you can also use `destructure!` in other places you might see a pattern:
+
+```rust
+let destructure!(Foo::A { a }) = x else {
+    panic!("not Foo::A")
+};
+
+match x {
+    destructure!(Foo::A { a }) => /* ... */,
+    Foo::B { .. } => panic!("unexpected Foo::B")
+}
+```
+
+Now, it makes sense that patterns in `destructure!` can be refutable. Even this can be okay, as long as only the top-level pattern is refutable. If we allow nested patterns to be refutable as well, we can construct the following scenario:
+
+```rust
+match y {
+    destructure!(Bar { a, b: 1 }) => /* ... */,
+    _ => panic!(".b != 1")
+}
+```
+
+Here, if we were to move fields out of `Bar` from left to right, we might move `a`, then check whether `b` matches `1`, and if it doesn't we already moved `a` out. 
```

Comment:
I’m not sure this disadvantage is real. Why does the move have to happen before fully matching? As far as I know, the existing semantics of patterns are compatible with the steps being:

1. Test the scrutinee against the pattern.
2. Execute match guard expressions (using temporary shared borrows).
3. If the pattern and the guard succeed, perform the moves and borrows specified by the pattern.

Reply:
This is indeed how rust works today and while it's not yet guaranteed behavior I expect this is what we'll guarantee once get around to it.